azure-kusto-data/azure/kusto/data/exceptions.py

```diff
@@ -57,6 +57,66 @@ def get_partial_results(self) -> Optional[Dict[str, Any]]:
         return self.kusto_response
 
 
+class OneApiError:
+    def __init__(self, code: str, message: str, type: str, description: str, context: dict, permanent: bool) -> None:
+        self.code = code
+        self.message = message
+        self.type = type
+        self.description = description
+        self.context = context
+        self.permanent = permanent
+
+    @staticmethod
+    def from_dict(obj: dict) -> "OneApiError":
+        code = obj["code"]
+        message = obj["message"]
+        type = obj["@type"]
+        description = obj["@message"]
+        context = obj["@context"]
+        permanent = obj["@permanent"]
+        return OneApiError(code, message, type, description, context, permanent)
+
+
+class KustoMultiApiError(KustoServiceError):
+    """
+    Represents a collection of standard API errors from kusto. Use `get_api_errors()` to retrieve more details.
+    """
+
+    def __init__(self, errors: List[dict]):
+        self.errors = KustoMultiApiError.parse_errors(errors)
+        messages = [error.description for error in self.errors]
+        super().__init__(messages[0] if len(self.errors) == 1 else messages)
+
+    def get_api_errors(self) -> List[OneApiError]:
+        return self.errors
+
+    @staticmethod
+    def parse_errors(errors: List[dict]) -> List[OneApiError]:
+        parsed_errors = []
+        for error_block in errors:
+            one_api_errors = error_block.get("OneApiErrors", None)
+            if not one_api_errors:
+                continue
+            for inner_error in one_api_errors:
+                error_dict = inner_error.get("error", None)
+                if error_dict:
+                    parsed_errors.append(OneApiError.from_dict(error_dict))
+        return parsed_errors
+
+
+class KustoApiError(KustoServiceError):
```

Comment:
I'm not sure about the term ApiError, why not use KustoServiceError directly (we can take it offline)

Reply:
KustoServiceError doesn't necessarily have its errors in the ApiError object format

Reply:
ok

Reply:
@AsafMah , Ariel has a point here. KustoServiceError should express server returned errors.
We should use KustoClientErrors in other cases.
Can you do a quick check to see how far this statement is from the current impl?

Reply:
Here's how we use KustoServiceError:
```python
  if status == 404:
            if payload:
                raise KustoServiceError("The ingestion endpoint does not exist. Please enable streaming ingestion on your cluster.", response) from exception

            raise KustoServiceError(f"The requested endpoint '{endpoint}' does not exist.", response) from exception

        if payload:
            message = f"An error occurred while trying to ingest: Status: {status}, Reason: {response.reason}, Text: {response_text}."
            if response_json:
                raise KustoApiError(response_json, message, response) from exception

            raise KustoServiceError(message, response) from exception

        if response_json:
            raise KustoApiError(response_json, http_response=response) from exception

        if response_text:
            raise KustoServiceError(response_text, response) from exception

        raise KustoServiceError("Server error response contains no data.", response) from exception
```

As you can see, it's also for errors like 404 or data we can't parse,
which is why APIerror is a subclass.
Do you think we should change the type of these errors?

Reply:
I think that's ok as long it matches the .NET SDK (I thought it is a new term we are adding)